### ISSUE TITLE
simple pipeline config is slower on 1.5rc2 than 1.4.2

### ISSUE THREAD

**Issue by talevy** [elastic/logstash#2859]
config:
```ruby
input { 
  generator { 
    count => 2000000
    message => "."
  } 
}

output { 
stdout { codec => dots }
}
```

I get ~220kB/s throughput on 1.4.2 and ~140kB/s on 1.5 when measuring using `pv` tool

**Comment by talevy** [comment]
Rough benchmark test to find root of problem...

```ruby
# edited pipeline.rb for both 1.4.2 and 1.5
# somewhere inside pop & process loop within outputworker
puts Benchmark.measure { 100000.times {output(event) }}
```

logstash config:
```ruby
input { 
  generator { 
    count => 2000000
    message => "."
  } 
}

output { 
null {}
}
```

timing results (the sum of the user and system CPU times in seconds)
```
0.120000 # 1.4.2
0.180000 # 1.5
```

**Comment by jordansissel** [comment]
Testing on my laptop:

```
bin/logstash -e 'input { generator { count => 10000000 } } output { stdout { codec => dots } }' | pv -War > /dev/null
```

* 1.4.2 gets around 145k eps
* 1.5.0.beat1 gets around 108k eps

**Comment by suyograo** [comment]
@talevy thinks this is related to the generator input

**Comment by colinsurprenant** [comment]
So after some testing I found out that in **master**, running @jordansissel config using 

```sh
bin/logstash -e 'input { generator { count => 10000000 } } output { stdout { codec => dots } }' | pv -Wbar > /dev/null
````

is ~20% slower than running it with a **local JRuby with same vendored version 1.7.17** using:

```sh
USE_RUBY=1 bin/logstash -e 'input { generator { count => 10000000 } } output { stdout { codec => dots } }' | pv -Wbar > /dev/null
```

I just tried commenting out the `JAVA_OPTS` in https://github.com/elastic/logstash/blob/master/bin/logstash.lib.sh#L22-L28 and it seems to solve the performance issue. I haven't tried disabling the options one by one to see which one is problematic.

**Comment by colinsurprenant** [comment]
the culprit is
```sh
JAVA_OPTS="$JAVA_OPTS -XX:+UseConcMarkSweepGC"
```

**Comment by colinsurprenant** [comment]
We have to be careful how we measure performance when we say 1.4 vs 1.5, especially when running from a repo clone because I have noticed that by default, running 1.4.2 `bin/logstash` uses the local `ruby` command and not it's vendored JRuby, which means that the `JAVA_OPTS` are not set, in other words, bare 1.5 `bin/logstash` uses vendored JRuby + `JAVA_OPTS` but bare 1.4.2 `bin/logstash` uses local `ruby`, no `JAVA_OPTS` - so this will lead to a 20% performance drop for the 1.5 tests. 

To make sure to run with the same environment:

- 1.4.2

by default the local `ruby` is used. use `USE_RUBY=1` and `USE_JRUBY=1` to control.
```sh
# use locally installed ruby
USE_RUBY=1 bin/logstash ...

# use vendored JRuby
USE_JRUBY=1 bin/logstash ...
```

- 1.5

by default the vendored `JRuby` is used. use `USE_RUBY=1` to control.
```sh
# use locally installed ruby
USE_RUBY=1 bin/logstash ...

# use vendored JRuby
bin/logstash ...
```

**Comment by colinsurprenant** [comment]
to wrap this up:
- we identified a slowdown introduced by the `-XX:+UseConcMarkSweepGC` `JAVA_OPTS` option which, depending on you setup will be set set by default when using the vendored JRuby. We realized that when starting 1.4.2 from a repo fork and with a local JRuby available, the local JRuby will be used and the `JAVA_OPTS` will not be set but in 1.5.0, the vendored JRuby and the default `JAVA_OPTS` will always be used by default. #2914 has been created for this.

- a few pipeline performance regressions have been fixed in #2869

I will go ahead and close this issue.
